File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/GLWHandler.java
package com.spvessel;

// import java.util.*;
// import java.io.FileReader;
// import java.io.BufferedReader;
// import java.util.ArrayList;
// import java.util.List;

// import Common.*;
// import Cores.*;
//import org.lwjgl.glfw.;
import com.spvessel.Core.Pointer;
import com.spvessel.Exceptions.SpaceVILException;
import org.lwjgl.glfw.*;
// import org.lwjgl.opengl.GL;
// import org.lwjgl.system.*;

import static org.lwjgl.glfw.GLFW.*;
// import static org.lwjgl.opengl.GL.createCapabilities;
import static org.lwjgl.system.MemoryUtil.*;
// import java.nio.ByteBuffer;


final class GLWHandler {
    // cursors

    long _arrow;
    long _input;
    long _hand;
    long _resize_h;
    long _resize_v;
    long _resize_all;
    ///////////////////////////////////////////////

    GLFWWindowSizeCallback resizeCallback;
    GLFWCursorPosCallback mouseMoveCallback;
    GLFWMouseButtonCallback mouseClickCallback;
    GLFWScrollCallback mouseScrollCallback;
    GLFWWindowCloseCallback windowCloseCallback;
    GLFWWindowPosCallback windowPosCallback;
    GLFWKeyCallback keyPressCallback;
    GLFWCharModsCallback keyInputTextCallback;
    GLFWWindowFocusCallback windowFocusCallback;
    ///////////////////////////////////////////////

    protected Boolean borderHidden;
    protected Boolean appearInCenter;
    protected Boolean focusable;
    protected Boolean focused = true;
    protected Boolean resizeble;
    protected Boolean visible;
    protected Boolean alwaysOnTop;
    protected Boolean maximized;
    private Pointer wPosition = new Pointer();

    protected Pointer getPointer() {
        return wPosition;
    }
    ///////////////////////////////////////////////

    private WindowLayout _w_layout;

    protected WindowLayout getLayout() {
        return _w_layout;
    }

    long _window = NULL;

    protected long getWindowId() {
        return _window;
    }

    protected int gVAO = 0;

    protected GLWHandler(WindowLayout handler) {
        _w_layout = handler;
        getPointer().setX(0);
        getPointer().setY(0);
    }

    protected void initGlfw() throws SpaceVILException {
        // path to c++ glfw3.dll (x32 or x64)
        // glfwConfigureNativesDirectory(AppDomain.CurrentDomain.BaseDirectory);
        if (!glfwInit()) {
            // System.err.println("GLFW initialization failed!");
            // System.exit(-1);
            throw new SpaceVILException("Init GLFW fail - " + getLayout().getWindowTitle());
        }

        // cursors
        _arrow = glfwCreateStandardCursor(GLFW_ARROW_CURSOR);
        _input = glfwCreateStandardCursor(GLFW_IBEAM_CURSOR);
        _hand = glfwCreateStandardCursor(GLFW_HAND_CURSOR);
        _resize_h = glfwCreateStandardCursor(GLFW_HRESIZE_CURSOR);
        _resize_v = glfwCreateStandardCursor(GLFW_VRESIZE_CURSOR);
        _resize_all = glfwCreateStandardCursor(GLFW_CROSSHAIR_CURSOR);
    }

    protected void createWindow() throws SpaceVILException {
        // important!!! may be the best combination of WINDOW HINTS!!!
        glfwDefaultWindowHints();
        glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GLFW_TRUE);
        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_COMPAT_PROFILE);
        glfwWindowHint(GLFW_SAMPLES, 8);
        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);

        if (resizeble)
            glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE);
        else
            glfwWindowHint(GLFW_RESIZABLE, GLFW_FALSE);

        if (!borderHidden)
            glfwWindowHint(GLFW_DECORATED, GLFW_TRUE);
        else
            glfwWindowHint(GLFW_DECORATED, GLFW_FALSE);

        if (focused)
            glfwWindowHint(GLFW_FOCUSED, GLFW_TRUE);
        else
            glfwWindowHint(GLFW_FOCUSED, GLFW_FALSE);

        if (alwaysOnTop)
            glfwWindowHint(GLFW_FLOATING, GLFW_TRUE);
        else
            glfwWindowHint(GLFW_FLOATING, GLFW_FALSE);

        if (maximized)
            glfwWindowHint(GLFW_MAXIMIZED, GLFW_TRUE);
        else
            glfwWindowHint(GLFW_MAXIMIZED, GLFW_FALSE);

        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);

        _window = glfwCreateWindow(_w_layout.getWidth(), _w_layout.getHeight(), _w_layout.getWindowTitle(), NULL, NULL);

        if (_window == NULL) {
            System.out.println("glfwCreateWindow fail");
            throw new SpaceVILException("Create window fails - " + getLayout().getWindowTitle());
        }
        glfwMakeContextCurrent(_window);

        long monitor = glfwGetPrimaryMonitor();
        GLFWVidMode vidmode = glfwGetVideoMode(monitor);
        int width = vidmode.width();
        int height = vidmode.height();

        if (appearInCenter) {
            getPointer().setX((width - _w_layout.getWidth()) / 2);
            getPointer().setY((height - _w_layout.getHeight()) / 2);
        } else {
            getPointer().setX(_w_layout.getX());
            getPointer().setY(_w_layout.getY());
        }
        glfwSetWindowSizeLimits(_window, _w_layout.getMinWidth(), _w_layout.getMinHeight(), _w_layout.getMaxWidth(),
                _w_layout.getMaxHeight());
        glfwSetWindowPos(_window, getPointer().getX(), getPointer().getY());
        if (visible)
            glfwShowWindow(_window);
    }

    protected void switchContext() {
        glfwMakeContextCurrent(0);
        glfwMakeContextCurrent(_window);
    }

    protected void clearEventsCallbacks() {
        mouseMoveCallback = null;
        mouseClickCallback = null;
        mouseScrollCallback = null;
        keyPressCallback = null;
        keyInputTextCallback = null;
        windowCloseCallback = null;
        windowPosCallback = null;
        windowFocusCallback = null;
        resizeCallback = null;
    }

    protected void setCursorType(int type) {
        switch (type) {
        case GLFW_ARROW_CURSOR:
            glfwSetCursor(_window, _arrow);
            break;
        case GLFW_IBEAM_CURSOR:
            glfwSetCursor(_window, _input);
            break;
        case GLFW_CROSSHAIR_CURSOR:
            glfwSetCursor(_window, _resize_all);
            break;
        case GLFW_HAND_CURSOR:
            glfwSetCursor(_window, _hand);
            break;
        case GLFW_HRESIZE_CURSOR:
            glfwSetCursor(_window, _resize_h);
            break;
        case GLFW_VRESIZE_CURSOR:
            glfwSetCursor(_window, _resize_v);
            break;
        default:
            glfwSetCursor(_window, _arrow);
            break;
        }
    }

    protected Boolean isClosing() {
        return glfwWindowShouldClose(_window);
    }

    protected void destroy() {
        glfwDestroyWindow(_window);
    }

    protected void swap() {
        glfwSwapBuffers(_window);
    }

    protected void setToClose() {
        if (_window != 0)
            glfwSetWindowShouldClose(_window, true);
        // try
        // {
        // glfwSetWindowShouldClose(_window, true);
        // }
        // catch(Exception ex)
        // {
        // System.out.println(ex.toString());
        // }
    }

    protected void setCallbackMouseMove(GLFWCursorPosCallback function) {
        mouseMoveCallback = function;
        glfwSetCursorPosCallback(_window, mouseMoveCallback);
    }

    protected void setCallbackMouseClick(GLFWMouseButtonCallback function) {
        mouseClickCallback = function;
        glfwSetMouseButtonCallback(_window, mouseClickCallback);
    }

    protected void setCallbackMouseScroll(GLFWScrollCallback function) {
        mouseScrollCallback = function;
        glfwSetScrollCallback(_window, mouseScrollCallback);
    }

    protected void setCallbackKeyPress(GLFWKeyCallback function) {
        keyPressCallback = function;
        glfwSetKeyCallback(_window, keyPressCallback);
    }

    protected void setCallbackTextInput(GLFWCharModsCallback function) {
        keyInputTextCallback = function;
        glfwSetCharModsCallback(_window, keyInputTextCallback);
    }

    protected void setCallbackClose(GLFWWindowCloseCallback function) {
        windowCloseCallback = function;
        glfwSetWindowCloseCallback(_window, windowCloseCallback);
    }

    protected void setCallbackPosition(GLFWWindowPosCallback function) {
        windowPosCallback = function;
        glfwSetWindowPosCallback(_window, windowPosCallback);
    }

    protected void setCallbackFocus(GLFWWindowFocusCallback function) {
        windowFocusCallback = function;
        glfwSetWindowFocusCallback(_window, windowFocusCallback);
    }

    protected void setCallbackResize(GLFWWindowSizeCallback function) {
        resizeCallback = function;
        glfwSetWindowSizeCallback(_window, resizeCallback);
    }

    protected void setHidden(Boolean value) {
        if (value)
            glfwHideWindow(_window);
        else
            glfwShowWindow(_window);
    }
}

File: spacevil_21052019/net_src/Engine/GLWHandler.cs
#define OS_LINUX 

using System;
using System.Text;
using Glfw3;
using SpaceVIL.Common;
using SpaceVIL.Core;

namespace SpaceVIL
{
    internal sealed class GLWHandler
    {
        // private float _scaleWidth = 1.0f;
        // private float _scaleHeight = 1.0f;
        // float[] GetDpiScale()
        // {
        //     return new float[] { _scaleWidth, _scaleHeight };
        // }
        private void SetDpiScale(float w, float h)
        {
            // _scaleWidth = w;
            // _scaleHeight = h;
            // Console.WriteLine(String.Format("{0:0.0} {1:0.0}", w, h));
            DisplayService.SetDisplayDpiScale(w);
            _coreWindow.SetDpiScale(w, h);
        }

        ///////////////////////////////////////////////
        Glfw.WindowSizeFunc ResizeCallback;
        Glfw.CursorPosFunc MouseMoveCallback;
        Glfw.MouseButtonFunc MouseClickCallback;
        Glfw.ScrollFunc MouseScrollCallback;
        Glfw.WindowCloseFunc WindowCloseCallback;
        Glfw.WindowPosFunc WindowPosCallback;
        Glfw.KeyFunc KeyPressCallback;
        Glfw.CharModsFunc KeyInputText;
        Glfw.WindowFocusFunc WindowFocusCallback;
        Glfw.FramebufferSizeFunc FramebufferCallback;
        Glfw.WindowRefreshFunc WindowRefreshCallback;
        Glfw.DropFunc DropCallback;
        ///////////////////////////////////////////////

        internal bool BorderHidden;
        internal bool AppearInCenter;
        internal bool Focusable;
        internal bool Focused = true;
        internal bool Resizeble;
        internal bool Visible;
        internal bool AlwaysOnTop;
        internal bool Maximized;
        internal bool Transparent;
        private Pointer WPosition = new Pointer();
        internal Pointer GetPointer()
        {
            return WPosition;
        }
        ///////////////////////////////////////////////

        private CoreWindow _coreWindow;
        internal CoreWindow GetCoreWindow()
        {
            return _coreWindow;
        }
        Glfw.Window _window;
        internal Glfw.Window GetWindowId()
        {
            return _window;
        }
        internal uint[] GVAO = new uint[1];

        internal GLWHandler() { }
        internal GLWHandler(CoreWindow handler)
        {
            _coreWindow = handler;
            WPosition.SetX(0);
            WPosition.SetY(0);
        }

        private Pointer defWindowPos = new Pointer(200, 50);

        internal void CreateWindow()
        {
            //important!!! may be the best combination of WINDOW HINTS!!!

            // Glfw.DefaultWindowHints();
            Glfw.WindowHint(Glfw.Hint.OpenglForwardCompat, true);
            Glfw.WindowHint(Glfw.Hint.OpenglProfile, Glfw.OpenGLProfile.Core);
            Glfw.WindowHint(Glfw.Hint.Samples, _coreWindow._msaa);
            Glfw.WindowHint(Glfw.Hint.ContextVersionMajor, 3);
            Glfw.WindowHint(Glfw.Hint.ContextVersionMinor, 3);

            Glfw.WindowHint(Glfw.Hint.Resizable, Resizeble);
            Glfw.WindowHint(Glfw.Hint.Decorated, !BorderHidden);//make borderless window
            Glfw.WindowHint(Glfw.Hint.Focused, Focused);
            Glfw.WindowHint(Glfw.Hint.Floating, AlwaysOnTop);
            Glfw.WindowHint(Glfw.Hint.Maximized, Maximized);
            Glfw.WindowHint(Glfw.Hint.Visible, false);
            Glfw.WindowHint(Glfw.Hint.TranspatentFramebuffer, Transparent);

            _window = Glfw.CreateWindow(_coreWindow.GetWidth(), _coreWindow.GetHeight(), _coreWindow.GetWindowTitle());

            if (!_window)
            {
                LogService.Log().LogText("Create window fail - " + GetCoreWindow().GetWindowTitle());
                throw new SpaceVILException("SpaceVILException: Create window fail - " + GetCoreWindow().GetWindowTitle());
            }
            Glfw.MakeContextCurrent(_window);

            int w, h;
            Glfw.GetFramebufferSize(_window, out w, out h);

            SetDpiScale((float)w / (float)_coreWindow.GetWidth(), (float)h / (float)_coreWindow.GetHeight());

            if (AppearInCenter)
            {
                GetPointer().SetX((Glfw.GetVideoMode(Glfw.GetPrimaryMonitor()).Width - _coreWindow.GetWidth()) / 2);
                GetPointer().SetY((Glfw.GetVideoMode(Glfw.GetPrimaryMonitor()).Height - _coreWindow.GetHeight()) / 2);
                // _coreWindow.SetX(WPosition.GetX());
                // _coreWindow.SetY(WPosition.GetY());
            }
            else
            {
                // WPosition.SetX(_coreWindow.GetX());
                // WPosition.SetY(_coreWindow.GetY());
                _coreWindow.SetX(defWindowPos.GetX());//200);
                _coreWindow.SetY(defWindowPos.GetY());//50);
                GetPointer().SetX(defWindowPos.GetX());//200);
                GetPointer().SetY(defWindowPos.GetY());//50);
            }
            Glfw.SetWindowSizeLimits(_window, _coreWindow.GetMinWidth(), _coreWindow.GetMinHeight(), _coreWindow.GetMaxWidth(), _coreWindow.GetMaxHeight());
            Glfw.SetWindowPos(_window, WPosition.GetX(), WPosition.GetY());

            if (_coreWindow.IsKeepAspectRatio)
                Glfw.SetWindowAspectRatio(_window, _coreWindow.RatioW, _coreWindow.RatioH);

            if (Visible)
                Glfw.ShowWindow(_window);
        }

        internal void SwitchContext()
        {
            Glfw.MakeContextCurrent(Glfw.Window.None);
            Glfw.MakeContextCurrent(_window);
        }

        internal void ClearEventsCallbacks()
        {
            MouseMoveCallback = null;
            MouseClickCallback = null;
            MouseScrollCallback = null;
            KeyPressCallback = null;
            KeyInputText = null;
            WindowCloseCallback = null;
            WindowPosCallback = null;
            WindowFocusCallback = null;
            ResizeCallback = null;
            FramebufferCallback = null;
            WindowRefreshCallback = null;
            DropCallback = null;
        }

        internal void SetCursorType(EmbeddedCursor type)
        {
            switch (type)
            {
                case EmbeddedCursor.Arrow:
                    Glfw.SetCursor(_window, CommonService.CursorArrow);
                    break;
                case EmbeddedCursor.IBeam:
                    Glfw.SetCursor(_window, CommonService.CursorInput);
                    break;
                case EmbeddedCursor.Crosshair:
                    Glfw.SetCursor(_window, CommonService.CursorResizeAll);
                    break;
                case EmbeddedCursor.Hand:
                    Glfw.SetCursor(_window, CommonService.CursorHand);
                    break;
                case EmbeddedCursor.ResizeX:
                    Glfw.SetCursor(_window, CommonService.CursorResizeH);
                    break;
                case EmbeddedCursor.ResizeY:
                    Glfw.SetCursor(_window, CommonService.CursorResizeV);
                    break;
                default:
                    Glfw.SetCursor(_window, CommonService.CursorArrow);
                    break;
            }
        }

        internal bool IsClosing()
        {
            return Glfw.WindowShouldClose(_window);
        }
        internal void Destroy()
        {
            Glfw.DestroyWindow(_window);
        }
        internal void Swap()
        {
            Glfw.SwapBuffers(_window);
        }
        internal void SetToClose()
        {
            Glfw.SetWindowShouldClose(_window, true);
        }
        
        internal void SetCallbackMouseMove(Glfw.CursorPosFunc function)
        {
            MouseMoveCallback = function;
            Glfw.SetCursorPosCallback(_window, MouseMoveCallback);
        }
        internal void SetCallbackMouseClick(Glfw.MouseButtonFunc function)
        {
            MouseClickCallback = function;
            Glfw.SetMouseButtonCallback(_window, MouseClickCallback);
        }
        internal void SetCallbackMouseScroll(Glfw.ScrollFunc function)
        {
            MouseScrollCallback = function;
            Glfw.SetScrollCallback(_window, MouseScrollCallback);
        }
        internal void SetCallbackKeyPress(Glfw.KeyFunc function)
        {
            KeyPressCallback = function;
            Glfw.SetKeyCallback(_window, KeyPressCallback);
        }
        internal void SetCallbackTextInput(Glfw.CharModsFunc function)
        {
            KeyInputText = function;
            Glfw.SetCharModsCallback(_window, KeyInputText);
        }
        internal void SetCallbackClose(Glfw.WindowCloseFunc function)
        {
            WindowCloseCallback = function;
            Glfw.SetWindowCloseCallback(_window, WindowCloseCallback);
        }
        internal void SetCallbackPosition(Glfw.WindowPosFunc function)
        {
            WindowPosCallback = function;
            Glfw.SetWindowPosCallback(_window, WindowPosCallback);
        }
        internal void SetCallbackFocus(Glfw.WindowFocusFunc function)
        {
            WindowFocusCallback = function;
            Glfw.SetWindowFocusCallback(_window, WindowFocusCallback);
        }
        internal void SetCallbackResize(Glfw.WindowSizeFunc function)
        {
            ResizeCallback = function;
            Glfw.SetWindowSizeCallback(_window, ResizeCallback);
        }
        internal void SetCallbackFramebuffer(Glfw.FramebufferSizeFunc function)
        {
            FramebufferCallback = function;
            Glfw.SetFramebufferSizeCallback(_window, FramebufferCallback);
        }
        internal void SetCallbackRefresh(Glfw.WindowRefreshFunc function)
        {
            WindowRefreshCallback = function;
            Glfw.SetWindowRefreshCallback(_window, WindowRefreshCallback);
        }
        internal void SetCallbackDrop(Glfw.DropFunc function)
        {
            DropCallback = function;
            Glfw.SetDropCallback(_window, DropCallback);
        }

        internal void SetOpacity(float level)
        {
            Glfw.SetWindowOpacity(_window, level);
        }

        internal void SetHidden(bool value)
        {
            if (value)
                Glfw.HideWindow(_window);
            else
                Glfw.ShowWindow(_window);
        }
    }
}